### ISSUE TITLE
Fatal error while merging backups

### ISSUE THREAD

**Issue by M-i-ech** [bepaald/signalbackup-tools#105]
Not sure what information is relevant. But error itself is asking to be reported to author, so I'm doing so here.

I have two backups associated with same phone number, due to device juggling they have significant overlap (part before juggling), but for recent period, there are differences. I tried to merge them and certain threads appear to consistently throw fatal errors. 1 thread in common part, 2 threads in divergent parts.

For each thread, error appears to be the same:
```
Error : After sqlite3_step(): UNIQUE constraint failed: recipient.phone
  Query: "INSERT INTO recipient (_id,uuid,phone,email,group_id,blocked,message_ringtone,message_vibrate,call_ringtone,call_vibrate,notification_channel,mute_until,color,seen_invite_reminder,default_subscription_id,message_expiration_time,registered,system_display_name,system_photo_uri,system_phone_label,system_contact_uri,profile_key,signal_profile_name,signal_profile_avatar,profile_sharing,unidentified_access_mode,force_sms_selection,system_phone_type,uuid_supported,username,system_info_pending,storage_service_key,dirty,profile_family_name,profile_joined_name,profile_key_credential,group_type,gv2_capability,last_profile_fetch,mention_setting,storage_proto,capabilities,last_gv1_migrate_reminder,last_session_reset,wallpaper,wallpaper_file,about,about_emoji,system_family_name,system_given_name,extras,groups_in_common,chat_colors,custom_chat_colors_id,badges,pni,distribution_list_id,needs_pni_signature,unregistered_timestamp,hidden,reporting_token,system_nickname) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"
```

and later in output:

```
ERROR Foreign key constraint violated. This will not end well, aborting.
      Please report this error to the program author.
------------------------------
| table   | parent    | fkid |
------------------------------
| message | recipient | 0    |
| thread  | recipient | 0    |
------------------------------
A fatal error occurred while trying to import thread 133. Aborting
```

Some searching through issue reports and answers directed me at undocumented command `--checkdbintegrity`, however testing either backup with this command produces no errors.

I have not tried unpacking and repacking databases yet.

Within Signal the threads do have associated contacts and phone numbers, in fact 2 divergent threads are just a notification that number associated with very old contact signed up into Signal, with no further, or prior (within Signal database), interaction from either side.

Trying to recreate either backup within Signal doesn't change anything, so it's not a matter of somehow damaged backup file.

I can reproduce errors and provide further logs, I also have access to both devices and both divergent databases within respective Signal installations if that could provide any useful information.

**Comment by bepaald** [comment]
Thanks for reporting!

The real error at least is the `Error : After sqlite3_step(): UNIQUE constraint failed`. The `Foreign key` error that follows is the result of the previous error. And that first error, I think, must be caused by the program not recognizing two contacts in the two backups are actually the same.

Not sure what is going on here, maybe there is a contact (maybe you) in the two backups with the same phone number but different uuid's? That's just a first guess, but there's certainly a bug here. I'll think about it and maybe give you some instructions to test. Do you think you would you be able to inspect the sqlite database after unpacking the backup file?

**Comment by M-i-ech** [comment]
Thanks for quick answer.

As a test, I tried merging backups in opposite direction.

It turned out I made one mistake in my report, there are 2 threads common to both backups that cause errors, no matter the direction of merging (threads 133 and 151, I know those IDs don't mean anything without database, but I'm adding them for reference and in case I post further logs) and each backup has 1 unique thread that also causes this error (154 in what I would like to call "first" database and 155 in "second" database). Common threads between databases end at 153, as in, 154 in "first" is different thread with different contact than 154 in "second".

I'm not familiar with sql and databases, but if doing it is not too complex I should be able to figure out how to inspect it. Is there anything specific I should be looking for? Or just try to compare threads causing error with other threads and look for differences?

**Comment by M-i-ech** [comment]
No, while processing offending threads that phrase does not appear in the log.

More importantly, I downloaded sqlite database browser and looked through the data, your guess about uuid-s was correct.

Thread 133 (common) is associated with recipient 19 who has some uuid in "first" and blank field in "second".

Thread 151 (common) is associated with recipient 93 who has two different uuid-s.

Threads 154 (first) and 155 (second) are associated with recipients who have uuid in one database but blank field in other.

Forgive if I use terminology wrong, but I believe right term would be that databases were effectively forked from common version, and for the most part, Signal must have been somehow prompted to assign uuid-s to contacts that didn't have one previously. Since I do have common ancestor backup, I unpacked and checked it too, only one offending recipient has uuid assigned in that version, the one who inexplicably has 2 different uuid-s in "first" and "second".

It would appear that I have bothered you needlessly. I haven't tried fixing discrepancies and repacking databases yet, but it would seem fault was within my backups. Am I correct in thinking that manually overwriting uuids to match and repacking databases should fix the issue? Thank you for your help and sorry for wasting your time.

**Comment by bepaald** [comment]
>It would appear that I have bothered you needlessly. I haven't tried fixing discrepancies and repacking databases yet, but it would seem fault was entirely within my backups. Am I correct in thinking that manually overwriting uuids to match and repacking databases should fix the issue? Thank you for your help and sorry for wasting your time.

Absolutely not. This is a bug in my program and not an error in your databases. you should not need to unpack and manually adjust the backups to use this function.

I have been working on a fix, I'm assuming you are using windows, could you test the attached updated version?

[signalbackup-tools_win.zip](https://github.com/bepaald/signalbackup-tools/files/11091365/signalbackup-tools_win.zip)

Look out for errors, and the expected `Found existing thread for this recipient` and (new) `No existing thread found in target database` messages on the threads you mentioned.

If you are not on Windows, you will have to wait a little while until I push this, because I have to run my tests first and they take a while.

I am on my way out, so I might not be able to reply for a few hours.

Thank you!

**Comment by bepaald** [comment]
Ok, I made a stupid mistake before. Here is a version you can test:

[signalbackup-tools_win.zip](https://github.com/bepaald/signalbackup-tools/files/11093542/signalbackup-tools_win.zip)

But, even if you see no errors in the output, don't put this one on your phone yet, there is still one thing I need to do (but it may take a few days, I'll update here when it's done). But I'd love to hear if you see any more errors in the output.

**Comment by M-i-ech** [comment]
I only got a moment to run your test version after you posted an edit, but I've decided to try it anyway and post results, just in case they are any useful.

Program handled common threads without errors, from what I see, overwriting uuid present in "base" database with the one being imported.

However, when it came to thread that didn't exist in "base", only in source, program outputted:

```
  No existing thread found in target database for this recipient, importing.
  updateRecipientIds (2)
  Mapping 313 -> 297 (up)
Importing statements from source table 'avatar_picker'...0 entries...
Importing statements from source table 'recipient'...1 entries...
Error : After sqlite3_step(): UNIQUE constraint failed: recipient.phone
  Query: "INSERT INTO recipient (_id,uuid,phone,email,group_id,blocked,message_ringtone,message_vibrate,call_ringtone,call_vibrate,notification_channel,mute_until,color,seen_invite_reminder,default_subscription_id,message_expiration_time,registered,system_display_name,system_photo_uri,system_phone_label,system_contact_uri,profile_key,signal_profile_name,signal_profile_avatar,profile_sharing,unidentified_access_mode,force_sms_selection,system_phone_type,uuid_supported,username,system_info_pending,storage_service_key,dirty,profile_family_name,profile_joined_name,profile_key_credential,group_type,gv2_capability,last_profile_fetch,mention_setting,storage_proto,capabilities,last_gv1_migrate_reminder,last_session_reset,wallpaper,wallpaper_file,about,about_emoji,system_family_name,system_given_name,extras,groups_in_common,chat_colors,custom_chat_colors_id,badges,pni,distribution_list_id,needs_pni_signature,unregistered_timestamp,hidden,reporting_token,system_nickname) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"
```

However, after cursory glance at resulting database (after unpacking) it would seem it that thread did get imported and as with common threads, existing recipient data got overwritten with imported one. I don't have time to look deeper, into messages proper, until tomorrow.

As for whether it's an error in databases or bug in program, I assumed that in situation like that, where there are very similar but not identical database entries (like in this case, where almost all recipient info is exactly the same except of uuid) an interactive user decision might be necessary. After all there IS a discrepancy between databases and as it turns out, it is not safe to assume that different uuid-s (not blank versus present, actually different values) are supposed to mean different recipients and I'm not sure if any assumptions can truly safely be made.

**Comment by M-i-ech** [comment]
Only just now I had time to run second version.

Just like previous, for recipient data present in both, it overwrites data in base recipient table with imported one. Threads appear to be merged correctly (with duplicated, or not messages depending on option ``--removedoubles``).

Lines stating that thread was or wasn't found appear as they should.


For practical reasons, I'd like to be done with merging and get back to using Signal as normal as soon as possible, so I'll go with manually adjusting uuid discrepancies, repacking, merging using tool version from few days ago (before recent changes, which as you say aren't done yet) and if no errors are found, I'll restore such backup.

However, I will keep both databases and if everything goes right, I'll have spare device on which I can test things. If it'll be of any use, I offer to run tests, both on those databases and restoring modified backups to spare device.

**Comment by bepaald** [comment]
> Only just now I had time to run second version.
> 
> Just like previous, for recipient data present in both, it overwrites data in base recipient table with imported one. Threads appear to be merged correctly (with duplicated, or not, messages depending on option `--removedoubles`).
> 
> Lines stating that thread was or wasn't found appear as they should.
> 
> For practical reasons, I'd like to be done with merging and get back to using Signal as normal as soon as possible, so I'll go with manually adjusting uuid discrepancies, repacking, merging using tool version from few days ago (before recent changes, which as you say aren't done yet) and if no errors are found, I'll restore such backup.
> 
> However, I will keep both databases and if everything goes right, I'll have spare device on which I can test things. If it'll be of any use, I offer to run tests, both on those databases and restoring modified backups to spare device.

I understand. I just pushed the fixes I had to the repo. I'm still not entirely sure how to handle the remaining (possible) problems. The effect of these fixes is probably the same as manually adjusting the uuid's in one of the databases, but you should also be aware that problems _could_ arise. While almost everywhere in the database recipients are referenced by `recipient._id`, there are some places where they are referenced by uuid directly. Most notably, in group-updates (for example members added, or removed), the group members are identified by uuid. If such a such a message is present, the app will try to translate it into a _'Name has joined/left the group'_ message, but it might not be able to find the name that corresponds to the uuid if it was changed in the recipient table. Maybe Signal will just silently fail, or display an incorrect message, but it might also crash completely, I don't know. I could write a function to adjust all uuid's in those messages to the new one, but that will not be easy. Also, the uuid in these messages is sent together with a profile-key which may also be changed. Lastly, I'm not sure if I'm aware of all the places the uuid is still used in the database.

Do you have any idea how the uuid's ended up like this? At first, I thought just installing Signal on a new phone (without transferring any backups) could cause a new uuid to be generated, but I have contacts who have done this for sure, but looking back at my old and current backups, all my contacts still have the same uuid as they did when they were first introduced years ago.

**Comment by M-i-ech** [comment]
I just finished uploading.

There were a whole lot of "name has joined/left the group" even for non-group threads and for recipients entirely unrelated to the problematic 4. There was no failure, I tried sending some test messages, which apparently got delivered, but I'm going to have to wait a while for response, as I ended up working on it until quite late.

I don't know exactly why uuid-s ended like that, but I can tell story why I ended with different databases and perhaps you can glimpse something from that.

I've been using Signal for years. Probably about 8 or so by now. Switched to new device in 2018, with auto backups set on. For some reason, auto backups stopped around November 2020 (let's call this version "zero"), but I didn't realise that until April 2022, when phone screen failed. I managed to extract backup, but to my horror, last 1.5 year was missing. Since I liked this device, I wanted it repaired, but sourcing genuine parts was a real pain, and using non-genuine screen quickly turned out to not be an option due to abysmal quality. It took me a few months to buy everything I needed. In that time, I was using different device as temporary replacement, and Signal conversations on that device picked up from November 2020 version of backup. Only in February 2023 I was finally able to fix original phone properly, at which point I discovered that while I was signed out of Signal, all messages from November 2020 to April 2022 were intact.

Backup which I was calling "first" is a backup from original device, with messages up to April 2022, backup which I was calling "second" is backup created on device used in interim, containing messages up to November 2020 and then messages from April 2022 up until March 2023 (I was still using replacement specifically for Signal until couple days ago).

I can tell more detailed story of each of problematic recipients if it's necessary.

**Comment by M-i-ech** [comment]
> There were group-update messages in non-group threads?

Yes, within Signal there were "x left a group" and "group updated" notifications in non-group threads, in thread overview, including old common threads. After opening those threads and backing up to overview, group notification is gone and thread overview correctly displays what it should (mostly either last message or "X is in Signal").

>Are they also in (one of) the source databases, or did that happen during merging?

I don't know if any of those incorrect notifications would appear after uploading original databases. I can try testing it on "replacement" device that I'm not using any more, if you think it would be useful for you.

> Does that also happen if you use my (current) program directly without manually editing the databases first?

I don't know, like above, I can test it, but most likely over the weekend.

>What exactly did you do to prepare the databases before merging, just UPDATE recipient SET uuid = '[NEW_UUID]' WHERE uuid = '[OLD_ID]'; in one of the databases, or was there more to it?

Sadly, I don't know sql, so I used much more crude method.

I found some DB browser with GUI and used it to manually change uuid-s in "second" database to match those in "first". Since there were only 4 recipients it was pretty easy, I spent much more time looking through other tables looking where else uuid-s might be used, I found such column only in Payments table. In Identities table I found "address" column which clearly holds uuid-s too, but neither of those 2 tables in any DB version had record same as any of those 4 uuid-s.

> the quote_mentions field uses the uuid.

I'm lucky then, that I use Signal mostly for direct and not group communications. That field is blank for all messages across entire database.
